Adjust controller to process one ingress at a time. Use per-item rate limiting workqueue.

munnerz opened the pull request:
This is in relation to https://github.com/jetstack/cert-manager/issues/407

From that issue:

> Based on the code in WatchEvents (https://github.com/jetstack/kube-lego/blob/master/pkg/kubelego/watch.go#L73) - it appears that whenever any Kubernetes Ingress resource is created, updated, or deleted, all Ingress resources are immediately scheduled for re-processing.
>
> Ingresses that already have a valid certificate will be skipped, but any user with a number of failing/invalid ingresses will make requests to LE APIs in an attempt to validate those ingresses.
>
> As part of those syncs, more updates will likely be made to ingresses, thus re-queuing these ingresses to be immediately reprocessed after the 'round' of processing ingresses fails.
>
> The good news is we do only process one Ingress resource at a time, which should reduce the hits to the API somewhat (this could be a lot worse otherwise).

This PR changes the usage of the workqueue to only process one ingress at a time. It also switches the workqueue to use the rate limiting interface instead of a plain workqueue. This allows us to exponentially backoff validation attempts on a per ingress basis.

~**NOTE**: I have not run e2e tests against this patch yet - I will update this PR with results once I have~

/cc @jsha @simonswine 

ref #328

munnerz commented on pkg/kubelego/watch.go:
```diff
@@ -27,13 +27,25 @@ func ingressWatchFunc(c *kubernetes.Clientset, ns string) func(options k8sMeta.L
+		}
+		kl.workQueue.AddRateLimited(key)
+	}
+	return nil
```
This function is only used by the renewal ticker in order to trigger a resync of ingress resources.

Previously, it resynced all resources at once, with no kind of rate limit applied.

This changes that to add it via the rate limited queue. For items that are already failing, this will increment the rate limit slightly, however this is probably not an issue.

For the items the are currently not 'known' by the rate limiter (i.e. succeeding ingresses), they will be checked one by one after a delay of 10 minutes, which is also acceptable imo.

munnerz commented:
e2e tests passed after a few more commits 😃

munnerz commented on pkg/kubelego/watch.go:
```diff
@@ -49,10 +61,44 @@ func (kl *KubeLego) WatchReconfigure() {
-			kl.Log().Debugf("worker: done processing %v", item)
-			kl.workQueue.Done(item)
+			func(item interface{}) {
+				defer kl.workQueue.Done(item)
```
Yep - they serve two different purposes:

- *Done* will inform the queue that the particular work item has finished processing, as the workqueue will stop the same item being processed by two workers at once (although we actually only process one certificate at once anyway).

- *Forget* will forget the item altogether and clear its back off status.

So 'Done' should be called after after the function processing the results of Get() is done (regardless of err/success)

'Forget' should be called once the processing has 'succeeded' and we want to clear the rate limit state for that particular item.

Calling AddRateLimit along with Done will therefore *not* clear the rate limit. It will mark the queue as finished processing that particular item, and also queue the item to be re-added to the queue after the rate limit is up.

wallrj replied:
Looks like you have to call both:
```
	// Forget indicates that an item is finished being retried.  Doesn't matter whether its for perm failing
	// or for success, we'll stop the rate limiter from tracking it.  This only clears the `rateLimiter`, you
	// still have to call `Done` on the queue.
	Forget(item interface{})
```

wallrj commented on pkg/kubelego/watch.go:
```diff
@@ -27,13 +27,25 @@ func ingressWatchFunc(c *kubernetes.Clientset, ns string) func(options k8sMeta.L
 func (kl *KubeLego) WatchReconfigure() {
 
-	kl.workQueue = workqueue.New()
+	kl.workQueue = workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(time.Minute*10, time.Hour*24), "kube-lego")
```
10 minutes seems rather a long `basedelay`.
This means that kube-lego will send an API request to LetsEncrypt at most, once every ten minutes, right?
I'd have thought 1 API req/sec would be a reasonable base.
Maybe add a note to the docs explaining the new behaviour.
And do LE have any server side limits? 
If so, it'd cool if our LE client could understand HTTP 429 headers, does it already  do that?

munnerz replied:
So kube-lego attempts the entire validation flow in a single sync iteration, so each sync iteration can take up to 15 minutes at a time. We have a base back off of 10 minutes, which means after a single sync has failed, we'll wait 10 minutes before processing that item again.

munnerz replied:
Re 429 headers - these are sadly not yet returned 😢

wallrj commented on pkg/kubelego/watch.go:
```diff
@@ -49,10 +61,44 @@ func (kl *KubeLego) WatchReconfigure() {
+				}
+				name, namespace, err := cache.SplitMetaNamespaceKey(key)
+				if err != nil {
+					kl.Log().Errorf("worker: invalid string in workqueue: %s", item)
```
Maybe log the `err` here too.

wallrj commented on pkg/kubelego/watch.go:
```diff
@@ -49,10 +61,44 @@ func (kl *KubeLego) WatchReconfigure() {
+					// a rate limit is applied when adding the item after a failure
+					kl.workQueue.AddRateLimited(key)
+					return
+				}
```
Yeah , it looks like we should return the errors from `reconfigure`.

wallrj commented on pkg/kubelego/configure.go:
```diff
@@ -79,24 +76,13 @@ func (kl *KubeLego) processProvider(ings []kubelego.Ingress) (err error) {
+func (kl *KubeLego) reconfigure(ing kubelego.Ingress) error {
 	// setup providers
-	kl.processProvider(ingresses)
+	kl.processProvider(ing)
```
We're throwing away an `err` here.

wallrj commented on pkg/kubelego/watch.go:
```diff
@@ -61,7 +107,7 @@ func (kl *KubeLego) WatchEvents() {
 	kl.Log().Debugf("start watching ingress objects")
 
-	resyncPeriod := 60 * time.Second
+	resyncPeriod := 10 * time.Minute
```
Is this necessary? Setting this also to 10 minutes might cause confusion....this isn't what enforces the 1 req / 10 minute behaviour....the queue does that, so this being 1 minute shouldn't matter.
And as long as the informer and its API watch is working properly, this shouldn't need to be set so low.

wallrj commented on pkg/kubelego/watch.go:
```diff
@@ -70,15 +116,32 @@ func (kl *KubeLego) WatchEvents() {
 				return
 			}
 			kl.Log().Debugf("CREATE ingress/%s/%s", addIng.Namespace, addIng.Name)
-			kl.workQueue.Add(true)
```
Ok I get it. So previously we added a bool to the queue!?  which meant that all certificates were re-validated every time?

munnerz replied:
Yes spot on 😬

wallrj commented on pkg/kubelego/watch.go:
```diff
@@ -70,15 +116,32 @@ func (kl *KubeLego) WatchEvents() {
+				kl.Log().Infof("Detected deleted ingress %q - skipping", key)
+				// skip processing deleted items, as there is no reason to due to
+				// the way kube-lego serialises authorization attempts
+				// kl.workQueue.AddRateLimited(key)
```
Should we call `kl.workQueue.Forget(key)` here?

wallrj commented on pkg/kubelego/watch.go:
```diff
@@ -88,13 +151,24 @@ func (kl *KubeLego) WatchEvents() {
 			oldIng.ResourceVersion = ""
 			upIng.ResourceVersion = ""
```
These can be removed now.

wallrj commented on pkg/kubelego/watch.go:
```diff
@@ -88,13 +151,24 @@ func (kl *KubeLego) WatchEvents() {
 			oldIng.ResourceVersion = ""
 			upIng.ResourceVersion = ""
 
-			if !reflect.DeepEqual(oldIng, upIng) {
```
This would have played a large part of the flood of API requests I guess. If the status and meta were changing often.

munnerz replied:
Yes exactly! 😬

wallrj commented on pkg/kubelego/watch.go:
```diff
@@ -88,13 +151,24 @@ func (kl *KubeLego) WatchEvents() {
+			// we requeue ingresses only when their spec has changed, as the indicates
+			// a user has updated the specification of their ingress and as such we should
+			// re-trigger a validation if required.
+			if !reflect.DeepEqual(oldIng.Spec, upIng.Spec) {
```
Are annotations  / labels on the ingress object examined by Kube-lego? Should we compare those also?

munnerz replied:
Good point - if a user switches the `tls-acme` annotation from `false` to `true`, we won't notice it right now.

jsha commented:
Hi! Just checking in: Any ETA on responding to the above review and potentially merging & releasing? Thanks!

munnerz commented on pkg/kubelego/configure.go:
```diff
@@ -107,23 +93,11 @@ func (kl *KubeLego) reconfigure(ingressesAll []kubelego.Ingress) error {
 		kl.Log().Error("Error while processing certificate requests: ", strings.Join(errsStr, ", "))
-
-		// request a rerun of reconfigure
-		kl.workQueue.Add(true)
```
👍

munnerz commented on pkg/kubelego/configure.go:
```diff
@@ -79,24 +76,13 @@ func (kl *KubeLego) processProvider(ings []kubelego.Ingress) (err error) {
-	for _, ing := range ingressesAll {
-		if ing.Ignore() {
-			continue
-		}
```
👍

munnerz commented:
Hey @jsha - I've addressed the review comments. We're all at KubeCon this week so quite busy, but I will talk to @wallrj and try and get the latest commit re-reviewed.

As soon as this is merged, I'll then cut a new release of kube-lego.

jsha commented:
Excellent, thanks for the update!

wallrj commented on pkg/kubelego/configure.go:
```diff
@@ -53,32 +54,37 @@ func (kl *KubeLego) TlsIgnoreDuplicatedSecrets(tlsSlice []kubelego.Tls) []kubele
 		if err != nil {
-			provider.Log().Error(err)
-			continue
+			errs = append(errs, err)
```
:question:  Should we keep the `continue` ? We previously skipped the logic below when reset fails.

jsha commented:
Hi! Friendly post-KubeCon ping? This is still causing a lot of issues for us.

jsha commented:
Ping? :-)

jsha commented:
Thanks very much!
